File: 004_Median_of_Two_Sorted_Arrays.py
"""
There are two sorted arrays nums1 and nums2 of size m and n respectively.
Find the median of the two sorted arrays. The overall run time complexity
should be O(log (m+n)).
"""

# Key Point: Cuz time complexity should be within O(log M+N), we should
#            implement the algorithm by binary search.
#            Find median can be converted to find k-th smallest number.
#               index_short = min(len(short), )
#
# Special Case:
#            1) Two arrays separated ([1,2,3,4,5,6], [11,12,13,14,15])
#            2) Two arrays overlaped ([1,3,5,7,9,11,13], [2,4,6,8])
#            3) Equal numbers in arrays ([1,3,3,3,5,6], [2,3,3,3,3,4])
#            4) Empty array ([],[1,2,3]), ([1,2,3],[]), ([],[])
#
# Run time: 115ms / 139ms / 163 ms


class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: float
        """
        def binary_search(nums1, nums2, rank):
            # print nums1, nums2, rank
            # rank == 1, return the smallest element
            if not nums1:
                return nums2[rank-1]
            if not nums2:
                return nums1[rank-1]
            if rank == 1:
                return min(nums1[0], nums2[0])
            longer = nums1 if len(nums1)>=len(nums2) else nums2
            shorter = nums1 if len(nums1)<len(nums2) else nums2
            rank_s = min(len(shorter), rank/2)
            # the key idea of this algorithm is this 
            # rank_l + rank_s should always be rank
            # then we could always drop the smaller element
            # because the largest rank of the smaller element could
            # only be (rank - 1) !
            rank_l = rank - rank_s
            if longer[rank_l-1] == shorter[rank_s-1]:
                return longer[rank_l-1]
            elif longer[rank_l-1] < shorter[rank_s-1]:
                # kick the smaller num out of our search range
                return binary_search(longer[rank_l:], shorter[:rank_s], rank-rank_l)
            else:
                # kick the smaller num out of our search range
                return binary_search(shorter[rank_s:], longer[:rank_l], rank-rank_s)

            
        length = len(nums1) + len(nums2)
        if length & 1:
            ranks = [length/2 + 1]
        else:
            ranks = [length/2, length/2+1]
        for i,rank in enumerate(ranks):
            if i == 0:
                res = binary_search(nums1, nums2, rank)
            else:
                res = (res + binary_search(nums1, nums2, rank))/2.0
        return res

if __name__ == '__main__':
    sol = Solution()
    assert sol.findMedianSortedArrays([], [1,2,3,4,5]) == 3
    assert sol.findMedianSortedArrays([1], [2,3,4,5]) == 3
    assert sol.findMedianSortedArrays([1,2], [3,4,5]) == 3
    assert sol.findMedianSortedArrays([1,2,3], [4,5]) == 3
    assert sol.findMedianSortedArrays([1,2,3,4], [5]) == 3

    assert sol.findMedianSortedArrays([1,3], [2,4,5]) == 3
    assert sol.findMedianSortedArrays([1,4], [2,3,5]) == 3
    assert sol.findMedianSortedArrays([1,5], [2,3,4]) == 3

    assert sol.findMedianSortedArrays([2,3], [1,4,5]) == 3
    assert sol.findMedianSortedArrays([2,4], [1,3,5]) == 3
    assert sol.findMedianSortedArrays([2,5], [1,3,5]) == 3

    assert sol.findMedianSortedArrays([3,4], [1,2,5]) == 3
    assert sol.findMedianSortedArrays([3,5], [1,2,4]) == 3

    assert sol.findMedianSortedArrays([4,5], [1,2,3]) == 3
    assert sol.findMedianSortedArrays([1], [2,3,4,5]) == 3
    assert sol.findMedianSortedArrays([3], [1,2,4,5]) == 3

    assert sol.findMedianSortedArrays([1,2,3,4,5], [6,7,8,9,10]) == 5.5

File: c/004_Median_of_Two_Sorted_Arrays.c
/*There are two sorted arrays nums1 and nums2 of size m and n respectively. Find the median of the two sorted arrays. The overall run time complexity should be O(log (m+n)).*/


struct array_elem
{
	int val;
	int index;
};

double range_shrink(int* nums1, int* nums2, struct array_elem *num0, struct array_elem *num1, struct array_elem *num2, struct array_elem *num3, int totalSize);
int is_int(int num1, int dividend);

double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size) {

// array 1 [0...nums1Size-1];
// array 2 [0...nums2Size-1];

	struct array_elem range[4];
	struct array_elem temp;

	range[0].val = *nums1;
	range[0].index = 0;

	range[1].val = *(nums1+nums1Size-1);
	range[1].index = nums1Size-1;

	range[2].val = *nums2;
	range[2].index = 0;

	range[3].val = *(nums2+nums2Size-1);
	range[3].index = nums2Size-1;

	return range_shrink(nums1, nums2, &range[0], &range[1], &range[2], &range[3], nums1Size + nums2Size - 1);
}


//totalSize is m+n-1
double range_shrink(int* nums1, int* nums2, struct array_elem *num0, struct array_elem *num1, struct array_elem *num2, struct array_elem *num3, int totalSize)
{

	int mid_temp1;
	int mid_temp2;

	//base case: there is only 4 elements in the range
	if( (num0->index - num1->index <= 2)&&( num2->index - num3->index <= 2) )
	{
		num0->index = (num0->val < num2->val) ? (num2->index - 1 + num0->index):(num2->index + num0->index);
		num2->index = (num2->val < num0->val) ? (num0->index - 1 + num2->index):(num0->index + num2->index);

		if(num1->val > num3->val)
		{
			num1->index = num1->index + num3->index;
			if(num3->val < num0->val)
				num3->index = num0->index - 1 + num3->index;
			else
				num3->index = num0->index + num3->index;
		}
		else
		{
			num3->index = num1->index + num3->index;
			if(num1->val < num2->val)
				num1->index = num2->index - 1 + num1->index;
			else
				num1->index = num2->index + num1->index;
		}
		if(is_int(totalSize,2))
		{
			if(num0->index == totalSize/2)
				return num0->val;
			if(num1->index == totalSize/2)
				return num1->val;
			if(num2->index == totalSize/2)
				return num2->val;
			if(num3->index == totalSize/2)
				return (double)num3->val;
		}
		else
		{			
			if(num0->index == totalSize/2)
				mid_temp1 = num0->val;
			else if(num0->index == totalSize/2+1)
				mid_temp2 = num0->val;

			if(num1->index == totalSize/2)
				mid_temp1 = num1->val;
			else if(num1->index == totalSize/2+1)
				mid_temp2 = num1->val;

			if(num2->index == totalSize/2)
				mid_temp1 = num2->val;
			else if(num0->index == totalSize/2+1)
				mid_temp2 = num2->val;

			if(num3->index == totalSize/2)
				mid_temp1 = num3->val;
			else if(num0->index == totalSize/2+1)
				mid_temp2 = num3->val;

			return ((double)mid_temp1+(double)mid_temp2)/2;
		}
	}
	else
		return 0;
	//above is base case;

	

}

int is_int(int num, int dividend)
{
	if (num % dividend)
		return 0;
	else 
		return 1;
}

